### PR TITLE
Move OutOfMemory handling into the plasma store

### LINKED ISSUE #11772: [Object Spilling] Avoid client starvation when spilling
Currently, when spilling is triggered, clients retry indefinitely until space "shows up" in the object store.

As discovered in https://github.com/ray-project/ray/pull/11673, this simple policy doesn't work well when there are many concurrent clients spilling. Some clients might end up waiting forever for space to free up (e.g., if they are allocating larger objects, or get unlucky).

Solving this is a bit tricky given the single-threaded event loop of the plasma store. Ideally, we'd maintain a queue of client requests and process them in ~FIFO order for fairness. This might be possible to emulate without too large of a refactoring [requires design].

cc @stephanie-wang @rkooo567

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

This refactors the plasma store/client code to move all logic relating to creation requests into a queue centralized at the plasma store. OutOfMemory errors are now retried on a timer by the plasma store instead of the client and global GC is also triggered directly by the plasma store. Transient OutOfMemory errors (because the object store is full but objects are being spilled) are now retried on the same timer, but with an infinite number of retries and a short timeout of 100ms.

This also changes the behavior of OutOfMemory handling: previously clients would use exponential backoff up to some number of tries if the object store was full. Now the plasma store just retries after a set amount of time. We could still do the exponential backoff, but I don't think there's much of a benefit now that the clients no longer need to poll.

This also removes the callback that was used to trigger creation request handling after objects were spilled. I noticed that this callback actually still resulted in OOM errors because the spilled objects are asynchronously released after spilling is complete. Until we can make object unpinning from the NodeManager synchronous, we should just always retry on a timeout.

## Related issue number

Closes #11772.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stephanie-wang** [comment]
TODO: Unit test for the new creation queue class

File: src/ray/object_manager/plasma/store.h
Comment (ericl) [inline src/ray/object_manager/plasma/store.h:81]: Can we remove this and just have a small internal retry timeout of say `10ms`? Seems harmless to fire a timer at that frequency.

File: src/ray/object_manager/plasma/store.h
Comment (ericl) [inline src/ray/object_manager/plasma/store.h:132]: Is this TODO resolved now?

File: src/ray/object_manager/plasma/store.h
Comment (ericl) [inline src/ray/object_manager/plasma/store.h:112]: Can we remove these constants in favor of a hard-coded small timeout?

File: src/ray/object_manager/plasma/store.h
Comment (ericl) [inline src/ray/object_manager/plasma/store.h:372]: Comments

**Comment by ericl** [comment]
@stephanie-wang it looks pretty good, though when I try the following occasionally I still see OutOfMemory error:

```
(pid=raylet) [2020-11-10 16:44:24,254 E 32565 32579] store.cc:306: Not enough memory to create the object b175aa9d8c566fa3ffffffff0100000001000000, data_size=83886321, metadata_size=6, retrying after a timeout
(pid=raylet) [2020-11-10 16:44:25,255 E 32565 32579] store.cc:306: Not enough memory to create the object b175aa9d8c566fa3ffffffff0100000001000000, data_size=83886321, metadata_size=6, retrying after a timeout
(pid=raylet) [2020-11-10 16:44:26,255 E 32565 32579] store.cc:306: Not enough memory to create the object b175aa9d8c566fa3ffffffff0100000001000000, data_size=83886321, metadata_size=6, retrying after a timeout
(pid=raylet) [2020-11-10 16:44:27,255 E 32565 32579] store.cc:306: Not enough memory to create the object b175aa9d8c566fa3ffffffff0100000001000000, data_size=83886321, metadata_size=6, retrying after a timeout
(pid=raylet) [2020-11-10 16:44:28,256 E 32565 32579] store.cc:306: Not enough memory to create the object b175aa9d8c566fa3ffffffff0100000001000000, data_size=83886321, metadata_size=6, retrying after a timeout
(pid=raylet) [2020-11-10 16:44:29,256 E 32565 32579] store.cc:300: Not enough memory to create the object b175aa9d8c566fa3ffffffff0100000001000000, data_size=83886321, metadata_size=6, will send a reply of PlasmaError::OutOfMemory
```

Somehow this wasn't a transient retry?

```
import json
import time
import numpy as np
import ray

ray.init(
    num_cpus=1,
    object_store_memory=100 * 1024 * 1024,
    _system_config={
        "automatic_object_spilling_enabled": True,
        "max_io_workers": 100,
        "object_spilling_config": json.dumps(
            {"type": "filesystem", "params": {"directory_path": "/tmp/spill"}},
            separators=(",", ":")
        )
    },
)

@ray.remote
def f():
    return np.zeros(10 * 1024 * 1024)

ids = []
for _ in range(100):
    x = f.remote()
    ids.append(x)

#ray.wait(ids, num_returns=len(ids))
print("GET")
while ids:
    print(ray.get(ids.pop()))
```

Could we also add this as a python test?

**Comment by ericl** [comment]
Maybe the above problem is too complex to resolve without @rkooo567 's refactoring adding the ObjectCreationManager. Not sure why spill objects callback would fail to make enough space in the example though.

**Comment by rkooo567** [comment]
@ericl I probably know why (not 100% sure). Currently, when we decide `space_needed` within `SpillObjectsOfSize` method, we basically find the space needed from "pinned objects". But if we keep spilling objects, every object will be eventually unpinned (because we don't re-pin objects). So space_needed will return the wrong values

**Comment by rkooo567** [comment]
> Until we can make object unpinning from the NodeManager synchronous

Isn't this already happening? I can see we are unpinning objects inside `SpillObjectsOfSize` before calling RPCs. Maybe I am missing something here?

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (rkooo567) [inline src/ray/core_worker/store_provider/plasma_store_provider.cc:90]: Isn't this blocking call? If we cannot create the objects in a timely manner, the core worker will be blocked for a long time right?

File: src/ray/object_manager/plasma/store.h
Comment (rkooo567) [inline src/ray/object_manager/plasma/store.h:54]: TODO: unit test for this class => will you handle it in this PR? I am a little afraid of merging the PR without unit testing this component.

File: src/ray/object_manager/plasma/store.cc
Comment (rkooo567) [inline src/ray/object_manager/plasma/store.cc:1102]: Can we just use `asio_util.h::execute_after`?

File: src/ray/object_manager/plasma/store.cc
Comment (rkooo567) [inline src/ray/object_manager/plasma/store.cc:1142]: In this case, why don't we move the request to the end of the queue again? I guess that's the matching behavior from the previous impl, and also we can avoid starvation in this case? It seems like without doing this, we will wait without creating other queued objects until this object is actually created after all timeout?

**Comment by ericl** [comment]
@rkooo567 the explicit spill call and related problems go away with ObjectCreationManager right? Perhaps we can defer fixing this until then. It probably means spilling isn't ready for users to try out until then though.

File: src/ray/object_manager/plasma/store.cc
Comment (ericl) [reply src/ray/object_manager/plasma/store.cc:1142]: FIFO avoid starvation right?

Comment (rkooo567) [reply src/ray/object_manager/plasma/store.cc:1142]: It seems like after this error occurs, we don't pop out the first entry (which causes this error)? 

```cpp
  for (auto request_it = queue_.begin();
      request_it != queue_.end(); ) {
    if (!ProcessRequest(request_it->first, request_it->second, retry_after_ms)) {
      return false;
    }
```

In this case, doesn't it keep retrying until this object is creatable?

**Comment by rkooo567** [comment]
@ericl Ah, that's actually true. I think all these issues will be solved in that case.

**Comment by ericl** [comment]
Hmm, I think the criteria for OOM should be something like:

Given object size N.
OOM iff `N > total_plasma_store_size - total_mmapped_object_size`.

This condition should also be equivalent to "put failed even after all spillable objects have been spilled".

The current criteria is brittle since it doesn't prove we will hit OOM, just that we can't unpin any more objects.

**Comment by stephanie-wang** [comment]
Yeah, unfortunately this PR won't solve all the issues with OutOfMemory retries. Part of the reason is that it's still possible for memory to get held up even when objects are getting spilled. I've seen it in two ways:
1. Although the node manager synchronously unpins its plasma buffer, the release at the plasma store is asynchronous. This is because the plasma buffer destructor doesn't release the memory immediately; it sends an IPC to the plasma store to do the release.
2. The plasma store can have un-evictable memory for other reasons. The main one that I saw during testing was when one object was created, but not yet sealed or released. Then the next creation request in the queue would hit the intransient OOM condition.

Long-term, I think we need to get rid of the retries completely and rely on a single longer timer to allow time for global GC, etc. In the meantime, it's a bit too much for this PR, so that's why I have the two different timeouts. The short one is for transient OOM, when we know a local object will get released soon, and the longer one is for OOMs where we might be able to make space in the cluster.

We should also add a call to retry the creation requests each time the plasma store has released memory, so that we can serve a request as soon as there is memory available. That would also solve the first issue (unpinning from the plasma client is asynchronous). We could also try to track the memory more precisely, so that we always know for sure if the requests in the creation queue are serviceable.

File: src/ray/object_manager/plasma/store.h
Comment (stephanie-wang) [reply src/ray/object_manager/plasma/store.h:81]: See my comment on the main PR thread about why I have two timeouts.

I'll set the internal one to 10ms, but let me know what you think about having the two timeouts. I worry about using such a low timeout for the intransient OOM error.

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (stephanie-wang) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc:90]: I think this is fine since object creation is already a blocking call. I don't think this will block the IO event loop on the core worker.

File: src/ray/object_manager/plasma/store.cc
Comment (stephanie-wang) [reply src/ray/object_manager/plasma/store.cc:1142]: Yeah, all of the creation requests will be blocked until the first object gets created or fails due to OOM.

File: src/ray/object_manager/plasma/store.cc
Comment (stephanie-wang) [reply src/ray/object_manager/plasma/store.cc:134]: It is used?

Comment (rkooo567) [reply src/ray/object_manager/plasma/store.cc:134]: Sorry 😂 you are right. Please ignore this

File: src/ray/core_worker/store_provider/plasma_store_provider.cc
Comment (rkooo567) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc:90]: I see. I didn't know it was already a blocking call! But looking at code, this seems to still block the io event loop? Maybe something we should take a look in the future if I am not missing something here.

**Review by ericl — APPROVED** [review]

**Review by rkooo567 — APPROVED** [review]

**Comment by ericl** [comment]
=================================== FAILURES ===================================
_______________________ test_put_pins_object[157286400] ________________________
ray_start_object_store_memory = 157286400
    @pytest.mark.parametrize(

**Comment by stephanie-wang** [comment]
Blocked because we need to make Create requests async/polling to avoid blocking the core worker and raylet threads.

**Pull request closed without merging**
